mpt_trie/src/trie_ops.rs

```diff
@@ -6,13 +6,41 @@ use std::{fmt::Display, mem::size_of};
 use enum_as_inner::EnumAsInner;
 use ethereum_types::{H256, U128, U256, U512};
 use log::trace;
+use thiserror::Error;
 
 use crate::{
     nibbles::{Nibble, Nibbles},
     partial_trie::{Node, PartialTrie, WrappedNode},
     utils::TrieNodeType,
 };
 
+/// Stores the result of trie operations. Returns a [TrieOpError] upon
+/// failure.
+pub type TrieOpResult<T> = Result<T, TrieOpError>;
+
+/// An error type for trie operation.
+#[derive(Debug, Error)]
+pub enum TrieOpError {
+    /// An error that occurs when a hash node is found during an insert
+    /// operation.
+    #[error("Found a `Hash` node during an insert in a `PartialTrie`! These should not be able to be traversed during an insert!")]
+    HashNodeInsertError(),
+
+    /// An error that occurs when a hash node is found during a delete
+    /// operation.
+    #[error("Attempted to delete a value that ended up inside a hash node")]
+    HashNodeDeleteError(),
+
+    /// An error that occurs when a hash node is found during an extension node
+    /// collapse.
+    #[error("Extension managed to get a child node type that is impossible! (child: {0})")]
```

Comment:
I feel we should document what we mean by "impossible"

Reply:
paste what is meant by impossible in this case and will update the comment